File: backend/core/events.py
import json
from dataclasses import asdict, dataclass
from enum import Enum


class EventType(str, Enum):
    STAGE = "stage"
    PAPERS = "papers"
    GAPS = "gaps"
    TOKEN = "token"
    DONE = "done"
    ERROR = "error"


@dataclass
class SSEEvent:
    type: str
    content: str = ""
    stage: str = ""
    data: object = None


def fmt(event: SSEEvent) -> str:
    return f"data: {json.dumps(asdict(event), ensure_ascii=False)}\n\n"


def extract_token(raw_sse: str) -> str:
    try:
        ev = json.loads(raw_sse.removeprefix("data: ").strip())
        if ev.get("type") == EventType.TOKEN:
            return ev.get("content", "")
    except Exception:
        pass
    return ""

File: frontend/src/types/events.ts
export type EventType = 'stage' | 'papers' | 'gaps' | 'references' | 'editing_stage' | 'editing_patch' | 'editing_gate' | 'token' | 'done' | 'error'

export interface SSEEvent {
  type: EventType
  content?: string
  stage?: string
  data?: PaperItem[] | GapItem[] | ReferenceEventItem[]
}

export interface PaperItem {
  id: string
  title?: string
  year?: number
  score?: number
}

export interface GapItem {
  id: string
  description?: string
  severity?: 'high' | 'medium' | 'low'
  addressed_by?: number
  score?: number
}

export interface ReferenceEventItem {
  id: string
  title?: string
  year?: number
  score?: number
  source?: string
  url?: string
  excerpt?: string
}

export type MessageRole = 'user' | 'assistant'

export interface Message {
  id: string
  role: MessageRole
  content: string
  isStreaming: boolean
  timestamp: number
  scene?: string
}
